src/nvim/autocmd.c

```diff
@@ -2465,45 +2476,53 @@ bool autocmd_delete_id(int64_t id)
 // ===========================================================================
 //  AucmdExecutable Functions
 // ===========================================================================
-char *aucmd_exec_default_desc(AucmdExecutable acc)
+
+/// Generate a string description of a callback
+static char *aucmd_callback_to_string(Callback cb)
 {
-  size_t msglen = 100;
+  // NOTE: this function probably belongs in a helper
 
-  switch (acc.type) {
-  case CALLABLE_CB:
-    switch (acc.callable.cb.type) {
-    case kCallbackLua: {
-      char *msg = (char *)xmallocz(msglen);
-      snprintf(msg, msglen, "<Lua function %d>", acc.callable.cb.data.luaref);
-      return msg;
-    }
-    case kCallbackFuncref: {
-      // TODO(tjdevries): Is this enough space for this?
-      char *msg = (char *)xmallocz(msglen);
-      snprintf(msg, msglen, "<vim function: %s>", acc.callable.cb.data.funcref);
-      return msg;
-    }
-    case kCallbackPartial: {
-      char *msg = (char *)xmallocz(msglen);
-      snprintf(msg, msglen, "<vim partial: %s>", acc.callable.cb.data.partial->pt_name);
-      return msg;
-    }
-    default:
-      return NULL;
-    }
+  // FIXME: enabling dynamic memory will cause problems when called from `nvim_get_autocmds()`
+
+  switch (cb.type) {
+  case kCallbackLua: {
+    // size_t msglen = 100;
+    // char *msg = (char *)xmallocz(msglen);
+    // snprintf(msg, msglen, "<Lua: %d>", cb.data.luaref);
+    // return msg;
+    return "<lua function>";
+  }
+  case kCallbackFuncref: {
+    // TODO(tjdevries): Is this enough space for this?
+    // size_t msglen = 100;
+    // char *msg = (char *)xmallocz(msglen);
+    // snprintf(msg, msglen, "<vim function: %s>", cb.data.funcref);
+    // return msg;
+    return "<vim function>";
+  }
+  case kCallbackPartial: {
+    // size_t msglen = 100;
+    // char *msg = (char *)xmallocz(msglen);
+    // snprintf(msg, msglen, "<vim partial: %s>", cb.data.partial->pt_name);
+    // return msg;
+    return "<vim partial>";
+  }
   default:
-    return NULL;
+    return "";
   }
+  return "";
 }
 
+/// Generate a string description for the command/callback of an autocmd
 char *aucmd_exec_to_string(AutoCmd *ac, AucmdExecutable acc)
   FUNC_ATTR_PURE
 {
   switch (acc.type) {
   case CALLABLE_EX:
     return acc.callable.cmd;
-  case CALLABLE_CB:
-    return ac->desc;
```

Comment:
@tjdevries used the `ac->desc` to store this info, but since we want to separate the `command` from the `desc`, that would require separate memory management for the return value of this function, AFAICS.

- alt1 (current solution):  return the string for the command without any further validation, and without including anything about callbacks, except that they're a "function"
- alt2: create a new entry exec_str that's part of the AutoCmd struct, which has auto memory management, same as desc

Reply:
I'm not sure I understand your argument here, but I think you're overcomplicating things. It's fine to alloc memory in `aucmd_callback_to_string` and return it from this function. We just need to make sure every call site of `aucmd_exec_to_string` frees it (which is no problem).